C#: Roslyn-based stub generation

hvitved opened the pull request:
This PR introduced a Roslyn-based stub generator, based on the existing (but much much slower) [QL based stub generator](https://github.com/github/codeql/blob/8e864ab84a8534d560d501514032ea3e7770ec3f/csharp/ql/src/Stubs/Stubs.qll). The `make_stubs_nuget.py` script is updated to use the new generator, and this PR updates some of the existing stubs with newly generated stubs.

Commit-by-commit review is suggested.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubGenerator.cs:
```diff
@@ -0,0 +1,84 @@
+            writer.WriteLine("// This file contains auto-generated code.");
+            writer.WriteLine($"// Generated from `{assembly.Identity}`.");
+
+            var visitor = new StubVisitor(assembly, writer);
+
+            visitor.StubAttributes(assembly.GetAttributes(), "assembly: ");
+
+            foreach (var module in assembly.Modules)
+            {
+                module.GlobalNamespace.Accept(new StubVisitor(assembly, writer));
+            }
+        }
```
Why do we create multiple `StubVisitor` instances?

tamasvajk commented on csharp/extractor/Semmle.Util/MemoizedFunc.cs:
```diff
@@ -0,0 +1,48 @@
+        this.f = f;
+    }
+
+    public T2 this[T1 s]
```
Nit: `MemoizedFunc<>` doesn't behave as a `Func<>`. It behaves as a dictionary. I would change this indexer to an `Invoke` method, and then the call-site would match the `Func<>` closer: 
```c#
var f0 = new Func<int, int>(i=>i);
var f1 = new MemoizedFunc<int, int>(f0);

f0.Invoke(1);
f1.Invoke(1);
```

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs:
```diff
@@ -0,0 +1,835 @@
+            case Accessibility.Public:
+                stubWriter.Write("public ");
+                break;
+            case Accessibility.Protected or Accessibility.ProtectedOrInternal:
+                stubWriter.Write("protected ");
+                break;
+            case Accessibility.Internal:
+                stubWriter.Write("internal ");
+                break;
+            case Accessibility.ProtectedAndInternal or Accessibility.ProtectedOrInternal:
+                stubWriter.Write("protected internal ");
+                break;
```
Why is `Accessibility.ProtectedOrInternal` handled in two places?

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs:
```diff
@@ -0,0 +1,835 @@
+                }
+                else
+                {
+                    stubWriter.Write("throw null");
```
Does `throw null` work in an attribute argument?

hvitved replied:
Possibly not, but this is mostly dead code for now. We can make it work if/when we need to stub other attributes than `System.FlagsAttribute`.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs:
```diff
@@ -0,0 +1,835 @@
+                WriteCommaSep(c.Values, StubTypedConstant);
+                stubWriter.Write("}");
+                break;
+            default:
```
`TypedConstantKind.Type` is not handled, wouldn't simply `typeof(System.Object)` cover that case? Changes are that we don't care about the real type in the stubs.

hvitved replied:
Again, it is not really used right now.

tamasvajk reviewed (commented):
Looks good to me. I've added some comments/questions.

One general question: we're not explicitly ordering the printed members. Do we know if the roslyn provided order is stable? If we used alphabetical ordering, then we would probably match what the previous stub generator did, and then verifying the correctness of the newly generated stubs would be easy.

hvitved commented on csharp/extractor/Semmle.Extraction.CSharp.DependencyStubGenerator/Program.cs:
```diff
@@ -0,0 +1,9 @@
+using Semmle.Util.Logging;
+
+var logger = new ConsoleLogger(Verbosity.Info);
+using var dependencyManager = new DependencyManager(".", DependencyOptions.Default, logger);
```
I'll keep it as-is, because `logger` is also used further down.

hvitved commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubGenerator.cs:
```diff
@@ -0,0 +1,84 @@
+            "stubgenerator.dll",
+            null,
+            references.Select(tuple => tuple.Item1),
+            new CSharpCompilationOptions(OutputKind.ConsoleApplication, allowUnsafe: true));
```
I just followed what the standalone extractor does.

michaelnebel commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs:
```diff
@@ -0,0 +1,835 @@
+        if (!attributeAllowList.Contains(qualifiedName))
+            return;
+
+        stubWriter.Write($"[{prefix}{qualifiedName.AsSpan(0, @class.GetQualifiedName().Length - 9)}");
```
Does `GetQualifiedName` always postfix an attribute with `Attribute` (or does `-9` mean something else)?

hvitved replied:
It is because attribute classes usually (always?) end with `Attribute`, but I guess we may as well handle the case where it doesn't.

michaelnebel replied:
I think that is just a convention. It is such a strong convention that there is actually compiler support that allows
```csharp
public class MyTestAttribute : Attribute { }

[MyTest]
public class Test1 { }

[MyTestAttribute]
public class Test2 { }
```
However, if you were to write
```csharp
public class MyTest : Attribute { }

[MyTest]
public class Test1 { }

[MyTestAttribute]
public class Test2 { }
```
then you get a compile error only on the attribute usage on `Test2`.

michaelnebel reviewed (commented):
Approximately half way through. Will continue later today.
This looks really really cool @hvitved . Really great (and impressive) work!

Besides the inlined questions and comments:
- Maybe run DCA since we change existing production code.
- Is it possible to make some unit tests somehow of the StubVisitor?

hvitved commented:
> Looks good to me. I've added some comments/questions.
> 
> One general question: we're not explicitly ordering the printed members. Do we know if the roslyn provided order is stable? If we used alphabetical ordering, then we would probably match what the previous stub generator did, and then verifying the correctness of the newly generated stubs would be easy.

I have addressed your comments, and the members are now sorted alphabetically. I simply updated the original stub regeneration commits with the newly generated stubs, instead of adding extra commits.

tamasvajk commented on csharp/ql/test/resources/stubs/_frameworks/Microsoft.AspNetCore.App/Microsoft.AspNetCore.Authentication.Abstractions.cs:
```diff
@@ -178,9 +162,8 @@ public interface IAuthenticationSchemeProvider
                 System.Threading.Tasks.Task<Microsoft.AspNetCore.Authentication.AuthenticationScheme> GetSchemeAsync(string name);
                 void RemoveScheme(string name);
-                bool TryAddScheme(Microsoft.AspNetCore.Authentication.AuthenticationScheme scheme) => throw null;
+                virtual bool TryAddScheme(Microsoft.AspNetCore.Authentication.AuthenticationScheme scheme) => throw null;
```
Is this `virtual` correct?

hvitved replied:
Appears so: https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.iauthenticationschemeprovider.tryaddscheme?view=aspnetcore-7.0#microsoft-aspnetcore-authentication-iauthenticationschemeprovider-tryaddscheme(microsoft-aspnetcore-authentication-authenticationscheme)

tamasvajk commented on csharp/ql/test/resources/stubs/_frameworks/Microsoft.AspNetCore.App/Microsoft.AspNetCore.Authentication.Abstractions.cs:
```diff
@@ -189,22 +172,18 @@ public interface IAuthenticationService
             }
-
-            public interface IAuthenticationSignInHandler : Microsoft.AspNetCore.Authentication.IAuthenticationHandler, Microsoft.AspNetCore.Authentication.IAuthenticationSignOutHandler
+            public interface IAuthenticationSignInHandler : Microsoft.AspNetCore.Authentication.IAuthenticationSignOutHandler, Microsoft.AspNetCore.Authentication.IAuthenticationHandler
```
A newline seems to be missing here.

hvitved replied:
Odd, `dotnet format` should take care of that.

hvitved replied:
I will reorder the interfaces alphabetically though.

tamasvajk commented on csharp/ql/test/resources/stubs/_frameworks/Microsoft.AspNetCore.App/Microsoft.AspNetCore.Authentication.Core.cs:
```diff
@@ -10,17 +9,15 @@ namespace Authentication
                 public AuthenticationFeature() => throw null;
-                public Microsoft.AspNetCore.Http.PathString OriginalPath { get => throw null; set => throw null; }
-                public Microsoft.AspNetCore.Http.PathString OriginalPathBase { get => throw null; set => throw null; }
+                public Microsoft.AspNetCore.Http.PathString OriginalPath { get => throw null; set { } }
```
Could also be `set => throw null;` to match all other callable bodies.

hvitved replied:
I think I prefer to avoid `throw null` when possible.

tamasvajk replied:
I have no preference other than consistency. Should `void` returning methods also have an empty block body?

tamasvajk commented on csharp/ql/test/resources/stubs/_frameworks/Microsoft.AspNetCore.App/Microsoft.AspNetCore.Components.Authorization.cs:
```diff
@@ -14,61 +13,53 @@ public class AuthenticationState
                 public abstract class AuthenticationStateProvider
                 {
-                    public event Microsoft.AspNetCore.Components.Authorization.AuthenticationStateChangedHandler AuthenticationStateChanged;
+                    public event Microsoft.AspNetCore.Components.Authorization.AuthenticationStateChangedHandler AuthenticationStateChanged { add { } remove { } }
```
Could be `add => throw null; remove => throw null;`. Or could be omitted altogether as before.

hvitved replied:
I will remove them.

tamasvajk commented on csharp/ql/test/resources/stubs/_frameworks/Microsoft.AspNetCore.App/Microsoft.AspNetCore.Components.Forms.cs:
```diff
@@ -12,93 +11,83 @@ namespace Forms
-                    public void MarkAsUnmodified(Microsoft.AspNetCore.Components.Forms.FieldIdentifier fieldIdentifier) => throw null;
                     public object Model { get => throw null; }
-                    public void NotifyFieldChanged(Microsoft.AspNetCore.Components.Forms.FieldIdentifier fieldIdentifier) => throw null;
+                    public void NotifyFieldChanged(in Microsoft.AspNetCore.Components.Forms.FieldIdentifier fieldIdentifier) => throw null;
```
Looks like the `in` was missing before.

hvitved commented on csharp/extractor/Semmle.Extraction.CSharp/Entities/UserOperator.cs:
```diff
@@ -191,12 +90,8 @@ public static bool TryGetOperatorSymbol(string methodName, out string operatorNa
         /// <returns>The converted name.</returns>
         private static string OperatorSymbol(Context cx, IMethodSymbol method)
         {
-            if (method.ExplicitInterfaceImplementations.Any())
```
Yes, indeed.

tamasvajk commented on csharp/ql/test/resources/stubs/Antlr3.Runtime/3.5.1/Antlr3.Runtime.cs:
```diff
@@ -1,46 +1,48 @@
+            public static int InitialBufferSize;
             public virtual void Load(System.IO.TextReader r, int size, int readChunkSize) => throw null;
-            public const int ReadBufferSize = default;
+            public static int ReadBufferSize;
```
`const` fields became `static`.

michaelnebel replied:
Does the stub just need to be updated or does the issue still persist?

hvitved commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs:
```diff
@@ -0,0 +1,835 @@
+            symbol.GetNamespaceMembers().Any(IsRelevantNamespace));
+    }
+
+    private static bool IsNotPublic(Accessibility accessibility) =>
```
It is used once inverted, and four times directly, so I think to prefer as-is.

tamasvajk commented on csharp/ql/test/resources/stubs/_frameworks/Microsoft.NETCore.App/System.AppContext.cs:
```diff
@@ -0,0 +1,2 @@
+// This file contains auto-generated code.
```
Empty files could be skipped.

tamasvajk commented on csharp/ql/test/resources/stubs/_frameworks/Microsoft.NETCore.App/System.Runtime.InteropServices.cs:
```diff
@@ -1974,155 +1290,620 @@ public struct ManagedToUnmanagedIn
                 {
-                    public enum MessageSendFunction : int
+                    public static System.Runtime.InteropServices.GCHandle CreateReferenceTrackingHandle(object obj, out System.Span<nint> taggedMemory) => throw null;
+                    public static unsafe void Initialize(delegate* unmanaged<void> beginEndCallback, delegate* unmanaged<nint, int> isReferencedCallback, delegate* unmanaged<nint, void> trackedObjectEnteredFinalization, System.Runtime.InteropServices.ObjectiveC.ObjectiveCMarshal.UnhandledExceptionPropagationHandler unhandledExceptionPropagationHandler) => throw null;
```
Is `nint` correct here? It was `System.IntPtr` before. (And it shows up as `System.IntPtr` in the docs)

hvitved replied:
I believe `nint` and `System.IntPtr` is the same thing.

hvitved commented:
> Approximately half way through. Will continue later today.
> This looks really really cool @hvitved . Really great (and impressive) work!

Thanks :-)

I have (force) pushed changes to address your comments as well.

tamasvajk reviewed (commented):
> I have addressed your comments, and the members are now sorted alphabetically. I simply updated the original stub regeneration commits with the newly generated stubs, instead of adding extra commits.

Thanks. It looks like the original stub generator did a bit more than ordering alphabetically. ( I think it might have generated the full signatures in memory, and then order them before writing.) I think it would be difficult to repro this based on Roslyn, which is a shame, it would have been great to compare the original and new stubs. (BTW, we might not have a stable order for overloads either.)

I've added some comments where I found differences between the old and new stubs.

hvitved commented:
> Thanks. It looks like the original stub generator did a bit more than ordering alphabetically. ( I think it might have generated the full signatures in memory, and then order them before writing.) I think it would be difficult to repro this based on Roslyn, which is a shame, it would have been great to compare the original and new stubs. (BTW, we might not have a stable order for overloads either.)

Indeed, that is what the QL based stub generator does, and we very much don't want to first compute things in memory and then write to disk later.

michaelnebel commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs:
```diff
@@ -0,0 +1,835 @@
+                c.DeclaredAccessibility == Accessibility.ProtectedOrInternal ||
+                containingTypes.Contains(c.ContainingType)
+            ).
+            OrderBy(c => c.Parameters.Length).FirstOrDefault();
```
Avoid sorting to get max:
```csharp
.MaxBy(c => c.Parameters.Length);
```

hvitved replied:
TIL

hvitved replied:
(Should be `MinBy`.)

michaelnebel replied:
Ah :-) 50/50 shot :-)

michaelnebel commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs:
```diff
@@ -0,0 +1,835 @@
+            stubWriter.Write("unsafe ");
+        }
+
+        if (explicitInterfaceSymbol is null && symbol.DeclaredAccessibility == Accessibility.Private)
```
Maybe a comment to explain the logic behind this.

hvitved replied:
Actually appears to be no longer needed.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs:
```diff
@@ -246,18 +247,13 @@ private void StubEvent(IEventSymbol symbol, IEventSymbol? explicitInterfaceSymbo
-        if (symbol.RemoveMethod is not null)
-            stubWriter.Write("remove {} ");
-        stubWriter.WriteLine("}");
+        StubWriter.Write(" {}");
```
I think you need to replace this with `StubWriter.Write(";");`

hvitved commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs:
```diff
@@ -0,0 +1,835 @@
+                    stubWriter.Write("in ");
+                    break;
+                default:
+                    stubWriter.Write($"/* TODO: {parameter.RefKind} */");
```
`In` and `RefReadOnly` appear to be the same thing.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubGenerator.cs:
```diff
@@ -56,13 +56,20 @@ private static void StubReference(ILogger logger, CSharpCompilation compilation,
+        visitor.StubWriter.WriteLine("// This file contains auto-generated code.");
+        visitor.StubWriter.WriteLine($"// Generated from `{assembly.Identity}`.");
 
         visitor.StubAttributes(assembly.GetAttributes(), "assembly: ");
```
I think we're not stubbing any assembly attributed at the moment, so this could be removed. Also, if we stubbed them, then the generated stubs would probably not compile due to each file containing assembly attributes.

hvitved commented on csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs:
```diff
@@ -830,4 +836,9 @@ public override void VisitProperty(IPropertySymbol symbol)
+
+    public void Dispose()
+    {
+        StubWriter.Dispose();
```
D'oh.

tamasvajk reviewed (approved):
LGTM. 
How much faster is this solution than the previous QL based one?

michaelnebel commented on csharp/ql/test/resources/stubs/Antlr3.Runtime/3.5.1/Antlr3.Runtime.cs:
```diff
@@ -1,46 +1,48 @@
-            public ANTLRStringStream(System.Char[] data, int numberOfActualCharsInArray) => throw null;
-            protected ANTLRStringStream() => throw null;
-            public virtual int CharPositionInLine { get => throw null; set => throw null; }
+            public virtual int CharPositionInLine { get => throw null; set { } }
```
`set` doesn't throw null?

hvitved replied:
No reason to, it is just a matter of making the stubs as simple as possible in terms of generated code.

michaelnebel commented:
> LGTM. How much faster is this solution than the previous QL based one?

Yeah, I am also really curious!

hvitved commented:
> How much faster is this solution than the previous QL based one?

It takes just a few seconds to generate stubs for the full standard library, so definitely much faster, but I don't have actual numbers from before.

michaelnebel reviewed (approved):
Great work Tom!
LGTM 👍 

Non-blocking question:
Are you planning on making it unit testable or introduce some integration like tests for? We do have tests for the queries used by the old stub generator.

hvitved commented:
Generating stubs for `Swashbuckle.AspNetCore.Swagger` (which includes `NetCore` and `AspNetCore`) takes 50 seconds for the QL based stub generator, while it takes 5 seconds for the new stub generator.

hvitved commented:
> Are you planning on making it unit testable or introduce some integration like tests for? We do have tests for the queries used by the old stub generator.

Unit tests, no. But I was thinking about adding an integration test.
